### ISSUE TITLE
`multiline` regex rules with inconsistent behaviour across engines

### ISSUE THREAD

**Issue by M4tteoP** [coreruleset/coreruleset#3277]
First report: https://github.com/coreruleset/coreruleset/issues/3266 (reporter: @leveryd)

### Describe the bug
Rules with the `rx` operator matching the "start of line" metacharacter (`^`) behave differently based on the WAF engine. It depends if the `multiline` modifier is enabled by default for each regex or not.
It leads to:
- Possible false positives
- Extra computational time for computing checks not intended by the rule(s) design.

### Default Modifiers
Not an expert on all the codebases, but to the best of what I understood:
#### ModSec2 (v2.9.7) (for Apache)
- Comes with `PCRE2_DOTALL | PCRE2_DOLLAR_ENDONLY` (see [here](https://github.com/SpiderLabs/ModSecurity/blob/v2.9.7/apache2/re_operators.c#L971)).
-  Under [`standalone/regex.c`](https://github.com/SpiderLabs/ModSecurity/blob/v2.9.7/standalone/regex.c#L75) there also is some code that includes `PCRE_MULTILINE` if `cflags` comes with `AP_REG_NEWLINE`. I guess it is at compilation time. But on this file, there is not even mention of PCRE2, which should have been added to 2.9.* 🧐

#### Libmodsecurity (v.3.0.9) (E.g. Nginx)
- Comes with `PCRE2_DOTALL|PCRE2_MULTILINE` (see [here](https://github.com/SpiderLabs/ModSecurity/blob/v3.0.9/src/utils/regex.cc#L68)).

#### Coraza (v3.0.3):
- Comes with `"(?sm)%s", options.Arguments`. Since [#732](https://github.com/corazawaf/coraza/pull/732), rx operator has been aligned to Libmodsecurity with both [dotall and multiline](https://github.com/corazawaf/coraza/blob/0dabae31d831f4f6d0f1bb57a1c5adc5cd4f2eb5/internal/operators/rx.go#L29).
- Notably without that PR `942522-7` test [New line bypass](https://github.com/coreruleset/coreruleset/blob/e36f27e1429a841e91996f4a521d40c996ec74eb/tests/regression/tests/REQUEST-942-APPLICATION-ATTACK-SQLI/942522.yaml#L107C18-L107C18) was failing. That being said, it could be just a matter of `DOTALL` and not `MULTILINE`. 

### Steps to reproduce
Example rule: `SecRule RESPONSE_BODY "@rx ^}" "id:911,phase:4, deny, status:403, capture,log, t:none, msg:'multiline works'"`
Request: `curl 'localhost:80/anything'`
Expected server response:
```
{
  "args": {},
  "data": "",
  "files": {},
  "form": {},
  "headers": {
    "Accept": "*/*",
    "Connection": "close",
    "Host": "localhost",
    "User-Agent": "curl/7.87.0"
  },
  "json": null,
  "method": "GET",
  "origin": "192.168.48.1",
  "url": "http://localhost/anything"
} 
```
Results:
- Apache: does `not` trigger this rule
- Nginx: `triggers` the rule, and writes the log, but can not enforce the deny (headers already sent)
- Caddy-Coraza: `triggers` the rule and denies the request

### Possible solutions
The goal is to achieve consistency across WAF Engines that supports CRS bringing their behavior in line with the reference platform (Apache) (see [August 2023 Monthly Chat Decisions](https://github.com/coreruleset/coreruleset/issues/3252#issuecomment-1668414215)).
Possible solutions are:

- Work with engine devs to align the default modifiers.
- Make regexes not dependent on the `multiline` modifier:
   - Exploring solutions provided at https://github.com/SpiderLabs/ModSecurity/issues/2921
   - Using `(?-m)` (@azurit's proposal).

**Comment by dune73** [comment]
Very good writeup, thanks (I edited it slightly for even better comprehension).

Question: I see the different behaviors of the engines. What are the reasons the ModSec 2.9 behavior is the preferred behavior (outside of being our standard). Or why is this better than the other variants?

**Comment by M4tteoP** [comment]
> What are the reasons the ModSec 2.9 behavior is the preferred behavior (outside of being our standard). Or why is this better than the other variants?

I think that the key point is to achieve default modifiers consistency across the engines, rather than push necessarily for one behavior or the other.

That being said, what I see in favor of ModSec 2.9 behavior (no `multiline` enabled by default) is:

- From a CRS standpoint, being ModSec 2.9 our standard, I think that rules have been written without the `multiline` enabled in mind. If we agree that it is actually needed, we have to rethink accordingly the rules to avoid unexpected matches.
- `Multiline` enabled for all the regex comes with some extra computational time, if it is something not broadly needed (It does not seem to be, considering that 2.9 never used it) I think it would be better to just enable it for specific rules that require it.

**Comment by theseion** [comment]
We do have a couple of rules that match newlines, those have definitely been written under the assumption that the documentation (https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual-%28v2.x%29#rx) is true and the input is a single line. From my perspective, there's no need for multiline, you can match the same thing as in the example with `[\r\n]+}`. Single line is also helpful in cases where you want to ignore lots of stuff and match something at the end. You normally don't know whether that's going to be on the same line or not, but in this case you can do `match1.*match2$`, for example. In a multiline setting (not sure whether `DOT_ALL` would match across multiple lines) you'd have use a chain rule, I think.

**Comment by dune73** [comment]
Why would you use the `+` in your example `[\r\n]+}`?

**Comment by theseion** [comment]
Oversight :) `[\r\n]}` does the job.

**Comment by dune73** [comment]
Thank you. 

(I do not want to sound nagging, but I want to be sure I really understand things.)

**Comment by theseion** [comment]
@M4tteoP to recap, IMO, it makes more sense for engines to compile with `DOT_ALL` and write rules that assume input on a single line.

**Comment by M4tteoP** [comment]
> it makes more sense for engines to compile with DOT_ALL and write rules that assume input on a single line.

This is the possible solution described as `Work with engine devs to align the default modifiers`.
To recap the current modifiers are:
| Engine | Modifier1 | Modifier2  |
| :---:   | :---: | :---: |
| Modsec v2 | PCRE2_DOTALL   | PCRE2_DOLLAR_ENDONLY   |
| Modsec v3 | PCRE2_DOTALL   | PCRE2_MULTILINE   |
| Coraza | s (dotall) | m (multiline mode)   |

The goal would be to align all of them to a **single and unified behavior**, and the preferred one should be the Modsec v2 behavior (No multiline).

**Coraza**: we had a discussion about this issue during our last meeting. The change is doable and seems legit. It is still not decided if we want to go for a breaking change for `v3.*` or make it a configurable option, and by default remove the multiline only from `v4.*`. Still, being the CRS support pretty important, even the breaking change is not excluded atm.
One note is that Coraza is running RE2, and I did not found a correspondence for `PCRE2_DOLLAR_ENDONLY`, but some tests are showing that the dollar endonly behavior is the default one for re2 🤔 See [here](https://github.com/corazawaf/coraza/pull/876/files#r1329315490).

**ModSecurity**: I don't have details ModSecurity-side.

**Comment by dune73** [comment]
Thank you. I think we need @airween to discuss the approach to get ModSec v3 to adopt the ModSec v2 standard of `PCRE2_DOLLAR_ENDONLY`: It's is the old standard, it's more performing, few rules use multiline, if anybody needs multiline it can be enabled on a case by case base within the regex.

**Comment by airween** [comment]
I reviewed the thread (congrats @M4tteoP for the good explanation), and tried the examples. I also checked the history of regex modifiers in v3, and realized unfortunately the "default" modifier is the `MULTIILINE` since the beginning.

I can make a PR, it's easily - but (because of the historical reason) I'm afraid that will be hard to accept.

**Comment by fzipi** [comment]
@airween what should we do next? Create issues for modsecurity from this one?

@M4tteoP I think coraza is already aligned, right?

**Comment by M4tteoP** [comment]
> @M4tteoP I think Coraza is already aligned, right?
There is an open PR that should still be ready to go: https://github.com/corazawaf/coraza/pull/876. The idea was to coordinate with ModSec and merge it once we have a single expected behaviour.

**Comment by fzipi** [comment]
Thanks Matteo. @airween ?

**Comment by airween** [comment]
> Thanks Matteo. @airween ?

I've opened an issue under ModSecurity, see [3295](https://github.com/owasp-modsecurity/ModSecurity/issues/3295)

**Comment by fzipi** [comment]
Status check on this — nothing has moved in about a year and a half, so here's where things actually stand:

- **ModSecurity v3 / libmodsecurity**: the tracking issue [owasp-modsecurity/ModSecurity#3295](https://github.com/owasp-modsecurity/ModSecurity/issues/3295) has had no activity since 2024-11-14. @airween, you mentioned you could put together a PR to align the default `@rx` flags to `PCRE2_DOTALL | PCRE2_DOLLAR_ENDONLY` (matching ModSec v2) — is that still something you're able to pick up, or should we look for someone else to drive it?
- **Coraza**: already did its half of the work in [corazawaf/coraza#876](https://github.com/corazawaf/coraza/pull/876) (merged 2024-11-18), but it landed as an opt-in build tag (`coraza.rule.no_regex_multiline`), not the default — the code even has a `// TODO: deprecate multiline modifier set by default in Coraza v4` comment, deliberately left non-default pending the ModSecurity v3 change.

So the remaining blocker is entirely on the ModSecurity v3 side. Until that PR exists, Coraza has no clear signal to flip its own default, and we're stuck with the three engines disagreeing on `^`/`$` semantics as described above.

---

**ai disclosure**: Claude (Sonnet 5) — assisted with: researching the current status of owasp-modsecurity/ModSecurity#3295 and corazawaf/coraza#876, drafting this status update. review performed: read Coraza's current `internal/operators/rx.go`/`multilineregex*.go` source directly to confirm its default is still non-multiline, and confirmed via the GitHub API that ModSecurity#3295 has had no activity since 2024-11-14.
